presto-spi/src/main/java/io/prestosql/spi/Page.java

```diff
@@ -23,35 +23,38 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
-import java.util.concurrent.atomic.AtomicLong;
 
 import static io.airlift.slice.SizeOf.sizeOf;
 import static io.prestosql.spi.block.DictionaryId.randomDictionaryId;
 import static java.lang.Math.min;
 import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
 
-public class Page
+public final class Page
 {
-    public static final int INSTANCE_SIZE = ClassLayout.parseClass(Page.class).instanceSize() +
-            (3 * ClassLayout.parseClass(AtomicLong.class).instanceSize());
+    public static final int INSTANCE_SIZE = ClassLayout.parseClass(Page.class).instanceSize();
 
     private final Block[] blocks;
     private final int positionCount;
-    private final AtomicLong sizeInBytes = new AtomicLong(-1);
-    private final AtomicLong retainedSizeInBytes = new AtomicLong(-1);
-    private final AtomicLong logicalSizeInBytes = new AtomicLong(-1);
+    private volatile long sizeInBytes = -1;
+    private volatile long retainedSizeInBytes = -1;
+    private volatile long logicalSizeInBytes = -1;
 
     public Page(Block... blocks)
     {
-        this(determinePositionCount(blocks), blocks);
+        this(true, determinePositionCount(blocks), blocks);
     }
 
     public Page(int positionCount, Block... blocks)
+    {
+        this(true, positionCount, blocks);
+    }
+
+    private Page(boolean blocksCopyRequired, int positionCount, Block[] blocks)
     {
         requireNonNull(blocks, "blocks is null");
-        this.blocks = Arrays.copyOf(blocks, blocks.length);
         this.positionCount = positionCount;
+        this.blocks = blocksCopyRequired ? blocks.clone() : blocks;
```

Comment:
does this bring any performance benefits? I think it's negligible optimization and can be removed

Reply:
In short, yes. This is proportional to the width of the page, so at a width of 256 columns without compressed oops- this is (256 * 8 bytes) ~2KiB for the array alone and that's essentially double since virtually all "new page" creating operations already allocated a Block array of that size before passing it in. This is often an "inner loop" to the real work being done on queries and workloads exists with significantly more than 256 columns. I'm actually more inclined to ask whether the defensive copy is carrying it's weight at all, but this iteration was just to remove it for "trusted" invocations of the constructor. Maybe a middle ground would be having a global flag (similar to "enable assertions") that could disable these copies entirely?